frame/screencopy_tracker.py

```diff
@@ -0,0 +1,133 @@
+    unlink_result = libc.shm_unlink(c_name)
+    assert unlink_result >= 0, f'Error {unlink_result} unlinking SHM file {name}: {os.strerror(ctypes.get_errno())}'
+    return open_result
+
+class ScreencopyTracker(WaylandClient):
+    required_extensions = ('zwlr_screencopy_manager_v1',)
+
+    def __init__(self, display_name: str) -> None:
+        super().__init__(display_name)
+        self.screencopy_manager: Optional[ZwlrScreencopyManagerV1Proxy] = None
+        self.output: Optional[WlOutputProxy] = None
+        self.shm: Optional[WlShmProxy] = None
+        self.frame: Optional[ZwlrScreencopyFrameV1Proxy] = None
+        self.buffer: Optional[WlBufferProxy] = None
+        self.shm_data: Optional[mmap.mmap] = None
+        self.frame_count = 0
+        self.total_damage = 0
+        self.buffer_width = 0
+        self.buffer_height = 0
+        self.buffer_size = 0
+        self.pending_damage = 0
+
+    def registry_global(self, registry, id_num: int, iface_name: str, version: int) -> None:
+        if iface_name == ZwlrScreencopyManagerV1.name:
+            self.screencopy_manager = registry.bind(id_num, ZwlrScreencopyManagerV1, version)
+        elif iface_name == WlOutput.name:
+            self.output = registry.bind(id_num, WlOutput, version)
+        elif iface_name == WlShm.name:
+            self.shm = shm = registry.bind(id_num, WlShm, version)
+
+    def connected(self) -> None:
+        self.copy_frame(True)
+
+    def disconnected(self) -> None:
+        if self.shm_data is not None:
+            self.shm_data.close()
+            self.shm_data = None
+
+    def _frame_buffer(self, frame, format: int, width: int, height: int, stride: int) -> None:
+        assert self.buffer_width == 0 or self.buffer_width == width, 'Buffer width changed'
+        self.buffer_width = width
+        assert self.buffer_height == 0 or self.buffer_height == height, 'Buffer height changed'
+        self.buffer_height = height
+        buffer_size = stride * height
+        assert self.buffer_size == 0 or self.buffer_size == buffer_size, 'Buffer size changed'
+        self.buffer_size = buffer_size
+        if self.buffer is None:
+            assert self.shm is not None, 'SHM not created'
+            fd = shm_open()
+            os.ftruncate(fd, buffer_size)
+            self.shm_data = mmap.mmap(fd, buffer_size)
+            shm_pool = self.shm.create_pool(fd, buffer_size)
+            libc.close(fd);
+            self.buffer = shm_pool.create_buffer(0, width, height, stride, format)
+            shm_pool.destroy()
+
+    def _frame_damage(self, frame, x: int, y: int, width: int, height: int) -> None:
+        self.pending_damage += width * height
+
+    def _frame_ready(self, frame, tv_sec_hi, tv_sec_lo, tv_nsec) -> None:
+        self.frame_count += 1
+        self.total_damage += (self.pending_damage
+            if self.pending_damage
+            else (self.buffer_width * self.buffer_height))
```

Comment:
```suggestion
        self.total_damage += self.pending_damage or (self.buffer_width * self.buffer_height)
```

Reply:
Perhaps too fancy. I think my way is too fancy too. I'm proposing we do it the slightly more verbose and much more obvious way:
```python
if self.pending_damage:
    self.total_damage += self.pending_damage
else:
    self.total_damage += self.buffer_width * self.buffer_height
```
No need to make someone scratch their head of two minutes to save three lines of code.

Reply:
Your call, the one-liner looks perfectly fine (aka _pythonic_) for me :)